File: Tesbo-Backend/src/main/java/com/bettercases/auth/OtpService.java
package com.bettercases.auth;

import com.bettercases.Config;
import com.bettercases.Database;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.SecureRandom;
import java.sql.*;
import java.time.Instant;
import java.util.Base64;
import java.util.Optional;
import java.util.UUID;

public final class OtpService {
    private static final SecureRandom RANDOM = new SecureRandom();
    private static final int OTP_LENGTH = 6;
    private final EmailService emailService = new EmailService();

    /** Returns true if OTP was created and email delivery was attempted. */
    public boolean requestOtp(String email, String ipAddress, String userAgent) {
        if (email == null || (email = email.trim().toLowerCase()).isEmpty()) return false;
        String ipKey = rateLimitKeyForIp(ipAddress);
        if (isRateLimited(email) || isRateLimited(ipKey)) return false;

        String plainCode = generateOtp();
        String codeHash = hash(plainCode);
        Instant expiresAt = Instant.now().plusSeconds(Config.OTP_EXPIRY_MINUTES * 60L);

        String sql = "INSERT INTO otp_codes (email, code_hash, expires_at) VALUES (?, ?, ?)";
        try (Connection c = Database.getDataSource().getConnection();
             PreparedStatement ps = c.prepareStatement(sql)) {
            ps.setString(1, email);
            ps.setString(2, codeHash);
            ps.setTimestamp(3, Timestamp.from(expiresAt));
            ps.executeUpdate();
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }

        recordOtpAttempt(email);
        recordOtpAttempt(ipKey);
        emailService.sendOtp(email, plainCode);
        return true;
    }

    /** Verifies OTP; returns session token if valid. OTP is single-use. */
    public Optional<String> verifyOtp(String email, String code, String ipAddress, String userAgent) {
        if (email == null || code == null) return Optional.empty();
        email = email.trim().toLowerCase();
        String ipKey = rateLimitKeyForIp(ipAddress);
        if (isRateLimited(email) || isRateLimited(ipKey)) return Optional.empty();

        String trimmedCode = code.trim();
        String codeHash = hash(trimmedCode);
        String selectSql = "SELECT id FROM otp_codes WHERE email = ? AND code_hash = ? AND expires_at > ? AND used_at IS NULL ORDER BY created_at DESC LIMIT 1";
        UUID otpId = null;
        try (Connection c = Database.getDataSource().getConnection();
             PreparedStatement ps = c.prepareStatement(selectSql)) {
            ps.setString(1, email);
            ps.setString(2, codeHash);
            ps.setTimestamp(3, Timestamp.from(Instant.now()));
            ResultSet rs = ps.executeQuery();
            if (rs.next()) otpId = (UUID) rs.getObject("id");
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }

        if (otpId == null) {
            recordOtpAttempt(email);
            recordOtpAttempt(ipKey);
            return Optional.empty();
        }

        markOtpUsed(otpId);
        Optional<UUID> userId = findOrCreateUser(email);
        if (userId.isEmpty()) return Optional.empty();
        String sessionToken = createSession(userId.get(), ipAddress, userAgent);
        clearRateLimit(email);
        clearRateLimit(ipKey);
        return Optional.of(sessionToken);
    }

    private boolean isRateLimited(String email) {
        String sql = "SELECT locked_until FROM otp_rate_limit WHERE email = ?";
        try (Connection c = Database.getDataSource().getConnection();
             PreparedStatement ps = c.prepareStatement(sql)) {
            ps.setString(1, email);
            ResultSet rs = ps.executeQuery();
            if (rs.next()) {
                Timestamp locked = rs.getTimestamp("locked_until");
                if (locked != null && locked.toInstant().isAfter(Instant.now()))
                    return true;
            }
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
        return false;
    }

    private void recordOtpAttempt(String key) {
        String sql = "INSERT INTO otp_rate_limit (email, attempt_count, locked_until, updated_at) VALUES (?, 1, ?, now()) " +
                "ON CONFLICT (email) DO UPDATE SET attempt_count = otp_rate_limit.attempt_count + 1, " +
                "locked_until = CASE WHEN otp_rate_limit.attempt_count + 1 >= ? THEN now() + (? || ' minutes')::interval ELSE otp_rate_limit.locked_until END, updated_at = now()";
        try (Connection c = Database.getDataSource().getConnection();
             PreparedStatement ps = c.prepareStatement(sql)) {
            ps.setString(1, key);
            ps.setNull(2, Types.TIMESTAMP);
            ps.setInt(3, Config.OTP_MAX_ATTEMPTS);
            ps.setInt(4, Config.OTP_RATE_LIMIT_WINDOW_MINUTES);
            ps.executeUpdate();
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    private void clearRateLimit(String email) {
        try (Connection c = Database.getDataSource().getConnection();
             PreparedStatement ps = c.prepareStatement("DELETE FROM otp_rate_limit WHERE email = ?")) {
            ps.setString(1, email);
            ps.executeUpdate();
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    private void markOtpUsed(UUID otpId) {
        try (Connection c = Database.getDataSource().getConnection();
             PreparedStatement ps = c.prepareStatement("UPDATE otp_codes SET used_at = now() WHERE id = ?")) {
            ps.setObject(1, otpId);
            ps.executeUpdate();
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    private Optional<UUID> findOrCreateUser(String email) {
        String selectSql = "SELECT id FROM users WHERE email = ?";
        try (Connection c = Database.getDataSource().getConnection();
             PreparedStatement ps = c.prepareStatement(selectSql)) {
            ps.setString(1, email);
            ResultSet rs = ps.executeQuery();
            if (rs.next()) return Optional.of((UUID) rs.getObject("id"));
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
        String insertSql = "INSERT INTO users (email, name) VALUES (?, ?) ON CONFLICT (email) DO NOTHING RETURNING id";
        try (Connection c = Database.getDataSource().getConnection();
             PreparedStatement ps = c.prepareStatement(insertSql)) {
            ps.setString(1, email);
            ps.setString(2, email.split("@")[0]);
            ResultSet rs = ps.executeQuery();
            if (rs.next()) return Optional.of((UUID) rs.getObject("id"));
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
        return findOrCreateUser(email);
    }

    private String createSession(UUID userId, String ipAddress, String userAgent) {
        byte[] bytes = new byte[32];
        RANDOM.nextBytes(bytes);
        String token = Base64.getUrlEncoder().withoutPadding().encodeToString(bytes);
        String tokenHash = hash(token);
        Instant expiresAt = Instant.now().plusSeconds(Config.SESSION_DAYS * 86400L);
        String sql = "INSERT INTO sessions (user_id, token_hash, user_agent, ip_address, expires_at) VALUES (?, ?, ?, ?, ?)";
        try (Connection c = Database.getDataSource().getConnection();
             PreparedStatement ps = c.prepareStatement(sql)) {
            ps.setObject(1, userId);
            ps.setString(2, tokenHash);
            ps.setString(3, userAgent);
            ps.setString(4, ipAddress);
            ps.setTimestamp(5, Timestamp.from(expiresAt));
            ps.executeUpdate();
            return token;
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    public Optional<UUID> resolveSession(String sessionToken) {
        if (sessionToken == null || sessionToken.isBlank()) return Optional.empty();
        String tokenHash = hash(sessionToken);
        String sql = "SELECT user_id FROM sessions WHERE token_hash = ? AND expires_at > ?";
        try (Connection c = Database.getDataSource().getConnection();
             PreparedStatement ps = c.prepareStatement(sql)) {
            ps.setString(1, tokenHash);
            ps.setTimestamp(2, Timestamp.from(Instant.now()));
            ResultSet rs = ps.executeQuery();
            if (rs.next()) return Optional.of((UUID) rs.getObject("user_id"));
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
        return Optional.empty();
    }

    public void invalidateSession(String sessionToken) {
        if (sessionToken == null || sessionToken.isBlank()) return;
        String tokenHash = hash(sessionToken);
        try (Connection c = Database.getDataSource().getConnection();
             PreparedStatement ps = c.prepareStatement("DELETE FROM sessions WHERE token_hash = ?")) {
            ps.setString(1, tokenHash);
            ps.executeUpdate();
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }

    private static String rateLimitKeyForIp(String ipAddress) {
        String normalizedIp = ipAddress == null ? "" : ipAddress.trim();
        return "ip:" + normalizedIp;
    }

    private static String generateOtp() {
        StringBuilder sb = new StringBuilder(OTP_LENGTH);
        for (int i = 0; i < OTP_LENGTH; i++) sb.append(RANDOM.nextInt(10));
        return sb.toString();
    }

    private static String hash(String value) {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            byte[] digest = md.digest(value.getBytes(StandardCharsets.UTF_8));
            return Base64.getUrlEncoder().withoutPadding().encodeToString(digest);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
}

File: Tesbo-Backend-Nest/src/auth/otp.service.ts
import { Injectable } from "@nestjs/common";
import { randomBytes, randomInt, createHash } from "crypto";
import { DatabaseService } from "../database/database.service";
import { AppConfigService } from "../config/app-config.service";
import { EmailService } from "./email.service";
import { SessionCacheService } from "../cache/session-cache.service";

@Injectable()
export class OtpService {
  constructor(
    private readonly db: DatabaseService,
    private readonly config: AppConfigService,
    private readonly email: EmailService,
    private readonly sessionCache: SessionCacheService
  ) {}

  async requestOtp(rawEmail: string, _ipAddress?: string | null, _userAgent?: string | null): Promise<boolean> {
    const email = rawEmail.trim().toLowerCase();
    if (!email) return false;

    const plainCode = this.generateOtp();
    const codeHash = this.hash(plainCode);
    const expiresAt = new Date(Date.now() + this.config.otpExpiryMinutes * 60_000);

    await this.db.query("INSERT INTO otp_codes (email, code_hash, expires_at) VALUES ($1, $2, $3)", [
      email,
      codeHash,
      expiresAt
    ]);
    await this.email.sendOtp(email, plainCode);
    return true;
  }

  async verifyOtp(rawEmail: string, code: string, ipAddress?: string | null, userAgent?: string | null): Promise<string | null> {
    const email = rawEmail.trim().toLowerCase();
    if (!(await this.verifyOtpCode(email, code))) return null;
    const userId = await this.findOrCreateUser(email);
    if (!userId) return null;
    return this.createSession(userId, ipAddress, userAgent);
  }

  async verifyOtpCode(rawEmail: string, code: string): Promise<boolean> {
    const email = rawEmail.trim().toLowerCase();

    const result = await this.db.query<{ id: string }>(
      "SELECT id FROM otp_codes WHERE email = $1 AND code_hash = $2 AND expires_at > now() AND used_at IS NULL ORDER BY created_at DESC LIMIT 1",
      [email, this.hash(code.trim())]
    );
    const otpId = result.rows[0]?.id;
    if (!otpId) return false;

    await this.markOtpUsed(otpId);
    return true;
  }

  async createSession(userId: string, ipAddress?: string | null, userAgent?: string | null): Promise<string> {
    const token = randomBytes(32).toString("base64url");
    const tokenHash = this.hash(token);
    const expiresAt = new Date(Date.now() + this.config.sessionDays * 86_400_000);
    await this.db.query(
      "INSERT INTO sessions (user_id, token_hash, user_agent, ip_address, expires_at) VALUES ($1, $2, $3, $4, $5)",
      [userId, tokenHash, userAgent ?? null, ipAddress ?? null, expiresAt]
    );
    return token;
  }

  async resolveSession(sessionToken: string): Promise<string | null> {
    if (!sessionToken?.trim()) return null;
    const tokenHash = this.hash(sessionToken);

    // Redis-backed cache, checked before Postgres: undefined = miss (fall through below), null =
    // known-invalid, string = the cached userId. A cache error already reads as undefined (miss), so
    // this is never worse than skipping the cache entirely.
    const cached = await this.sessionCache.get(tokenHash);
    if (cached !== undefined) return cached;

    const result = await this.db.query<{ user_id: string; expires_at: string }>(
      "SELECT user_id, expires_at FROM sessions WHERE token_hash = $1 AND expires_at > now()",
      [tokenHash]
    );
    const row = result.rows[0];
    if (!row) {
      await this.sessionCache.setInvalid(tokenHash);
      return null;
    }
    await this.sessionCache.setValid(tokenHash, row.user_id, new Date(row.expires_at));
    return row.user_id;
  }

  async invalidateSession(sessionToken: string): Promise<void> {
    if (!sessionToken?.trim()) return;
    const tokenHash = this.hash(sessionToken);
    await this.db.query("DELETE FROM sessions WHERE token_hash = $1", [tokenHash]);
    await this.sessionCache.invalidateToken(tokenHash);
  }

  /** Signs the user out of every session, including the one making this call. */
  async invalidateAllSessions(userId: string): Promise<void> {
    await this.db.query("DELETE FROM sessions WHERE user_id = $1", [userId]);
    await this.sessionCache.invalidateAllForUser(userId);
  }

  private async markOtpUsed(otpId: string): Promise<void> {
    await this.db.query("UPDATE otp_codes SET used_at = now() WHERE id = $1", [otpId]);
  }

  private async findOrCreateUser(email: string): Promise<string | null> {
    const existing = await this.db.query<{ id: string }>("SELECT id FROM users WHERE email = $1", [email]);
    if (existing.rows[0]) return existing.rows[0].id;
    const inserted = await this.db.query<{ id: string }>(
      "INSERT INTO users (email, name) VALUES ($1, $2) ON CONFLICT (email) DO NOTHING RETURNING id",
      [email, email.split("@")[0]]
    );
    return inserted.rows[0]?.id ?? (await this.findOrCreateUser(email));
  }

  private generateOtp(): string {
    return randomInt(0, 1_000_000).toString().padStart(6, "0");
  }

  private hash(value: string): string {
    return createHash("sha256").update(value, "utf8").digest("base64url");
  }
}
